mixing/mixpool/mixpool.go

```diff
@@ -0,0 +1,1523 @@
+func (p *Pool) HaveMessage(query *chainhash.Hash) bool {
+	p.mtx.RLock()
+	_, ok := p.pool[*query]
+	if !ok {
+		_, ok = p.prs[*query]
+	}
+	p.mtx.RUnlock()
+	return ok
+}
+
+// MixPR searches the mixing pool for a PR message by its hash.
+func (p *Pool) MixPR(query *chainhash.Hash) (*wire.MsgMixPairReq, error) {
+	var pr *wire.MsgMixPairReq
+
+	p.mtx.RLock()
+	pr = p.prs[*query]
+	p.mtx.RUnlock()
+
+	if pr == nil {
+		return nil, fmt.Errorf("PR message not found")
+	}
+
+	return pr, nil
+}
+
+// MixPRs returns all MixPR messages with hashes matching the query.  Unknown
+// messages are ignored.
+//
+// If query is nil, all PRs are returned.
+//
+// In both cases, any expired PRs that are still internally tracked by the
+// mixpool for ongoing sessions are excluded from the result set.
+func (p *Pool) MixPRs(query []chainhash.Hash) []*wire.MsgMixPairReq {
+	res := make([]*wire.MsgMixPairReq, 0, len(query))
+
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	p.removeConfirmedRuns()
+
+	if query == nil {
+		res = make([]*wire.MsgMixPairReq, 0, len(p.prs))
+		for _, pr := range p.prs {
+			// Exclude expired but not yet removed PRs.
+			if pr.Expiry <= p.expireHeight {
+				continue
+			}
+
+			res = append(res, pr)
+		}
+		return res
+	}
+
+	for i := range query {
+		pr, ok := p.prs[query[i]]
+		if ok {
+			// Exclude expired but not yet removed PRs.
+			if pr.Expiry <= p.expireHeight {
+				continue
+			}
+
+			res = append(res, pr)
+		}
+	}
+
+	return res
+}
+
+// CompatiblePRs returns all MixPR messages with pairing descriptions matching
+// the parameter.
+func (p *Pool) CompatiblePRs(pairing []byte) []*wire.MsgMixPairReq {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	res := make([]*wire.MsgMixPairReq, 0, len(p.prs))
+	for _, pr := range p.prs {
+		prPairing, _ := pr.Pairing()
+		if bytes.Equal(pairing, prPairing) {
+			res = append(res, pr)
+		}
+	}
+
+	// Sort by decreasing expiries and remove any PRs double spending an
+	// output with an earlier expiry.
+	sort.Slice(res, func(i, j int) bool {
+		return res[i].Expiry >= res[j].Expiry
+	})
+	seen := make(map[wire.OutPoint]uint32)
+	for i, pr := range res {
+		for _, utxo := range pr.UTXOs {
+			prevExpiry, ok := seen[utxo.OutPoint]
+			if !ok {
+				seen[utxo.OutPoint] = pr.Expiry
+			} else if pr.Expiry < prevExpiry {
+				res[i] = nil
+			}
+		}
+	}
+	filtered := res[:0]
+	for i := range res {
+		if res[i] != nil {
+			filtered = append(filtered, res[i])
+		}
+	}
+
+	// Sort again lexicographically by hash.
+	sort.Slice(filtered, func(i, j int) bool {
+		a := filtered[i].Hash()
+		b := filtered[j].Hash()
+		return bytes.Compare(a[:], b[:]) < 1
+	})
+	return filtered
+}
+
+// ExpireMessagesInBackground will, after the current epoch period ends,
+// remove all pair requests that indicate an expiry at or before the height
+// parameter and removes all messages that chain back to a removed pair
+// request.
+func (p *Pool) ExpireMessagesInBackground(height uint32) {
```

Comment:
It looks like this will deadlock if called more than once.

First call:
1) p.mtx is locked
2) Sends into p.expireSem which is a buffered chan of size 1
3) launches the goroutine
4) p.mtx is unlocked

Second call:
1) p.mtx is locked
2) Tries to send into `p.expireSem` which is blocked because it hasn't been read yet
3) meanwhile `expireMessages` goroutine is sleeping on `waitForExpiry`
4) `expireMessages` goroutine sleep ends and it tries to take `p.mtx`, but it's already locked.  Deadlock.

Reply:
this works because it's a non-blocking send

Reply:
this method is weird, and definitely up for changes later, but current behavior is commented better now.